tests/unit_tests/test_coverage.py

```diff
@@ -34,21 +89,116 @@ def test_coverage_get_from_spades(metagenome, tmp_path):
     assert out.exists()
 
 
-@pytest.mark.skip
-@pytest.mark.wip
-def test_coverage_get_from_sam(metagenome, tmp_path):
+def test_coverage_get_from_sam(metagenome, sam_alignment, tmp_path):
     out = tmp_path / "covs_from_sam.tsv"
-    df = coverage.get(fasta=metagenome, from_spades=False, out=out)
+
+    df = coverage.get(fasta=metagenome, from_spades=False, out=out, sam=sam_alignment)
     assert df.index.name == "contig"
     assert "coverage" in df.columns
     assert out.exists()
 
 
-@pytest.mark.skip
-@pytest.mark.wip
-def test_coverage_get_from_bam(metagenome, tmp_path):
+def test_coverage_get_from_bam(metagenome, bam_alignment, tmp_path):
     out = tmp_path / "covs_from_bam.tsv"
-    df = coverage.get(fasta=metagenome, from_spades=False, out=out)
+    df = coverage.get(fasta=metagenome, from_spades=False, out=out, bam=bam_alignment)
     assert df.index.name == "contig"
     assert "coverage" in df.columns
     assert out.exists()
+
+
+def test_coverage_get_from_bed(metagenome, bed_alignment, tmp_path):
+    out = tmp_path / "covs_from_bed.tsv"
+    df = coverage.get(
+        fasta=metagenome, from_spades=False, out=out.as_posix(), bed=bed_alignment
+    )
+    assert df.index.name == "contig"
+    assert "coverage" in df.columns
+    assert out.exists()
+
+
+def test_coverage_get_from_reads(metagenome, fwd_reads, rev_reads, tmp_path):
+    out = tmp_path / "covs_from_bed.tsv"
+    df = coverage.get(
+        fasta=metagenome,
+        from_spades=False,
+        out=out.as_posix(),
+        fwd_reads=fwd_reads,
+        rev_reads=rev_reads,
+    )
+    assert df.index.name == "contig"
+    assert "coverage" in df.columns
+    assert out.exists()
+
+
+def test_get_ValueError(metagenome, tmp_path):
+    out = tmp_path / "covs.tsv"
+    with pytest.raises(ValueError):
+        coverage.get(fasta=metagenome, from_spades=False, out=out)
+
+
+def test_embed_df_already_exists(metagenome, df_exists_fpath, bed_alignment):
+    coverage.get(fasta=metagenome, out=df_exists_fpath, bed=bed_alignment)
+
+
+use_spades = [True, False]
+
+
+@pytest.fixture(name="mock_parser", params=use_spades)
+def fixture_mock_parser(
+    metagenome,
+    sam_alignment,
+    bam_alignment,
+    bed_alignment,
+    fwd_reads,
+    rev_reads,
+    monkeypatch,
+    tmp_path,
+    request,
+):
+    def return_mock_parser(*args, **kwargs):
```

Comment:
As far as I can tell this does not need to be a fixture but can all be defined _within_ the `test_coverage_main` function, No?

Reply:
Also, there is no need to add all of these other fixtures to this. We just want to ensure the main logic is appropriate. We do _not_ want to rerun all of the functions together since we already have separate tests for them. This may be adding unnecessary time to our tests.